OvmfPkg/IoMmuDxe/AmdSevIoMmu.c

```diff
@@ -68,12 +69,371 @@ typedef struct {
+ *
+ * @param Type                Allocate type
+ * @param MemoryType          The memory type to be allocated
+ * @param Pages               Pages to be allocated.
+ * @param ReservedMemBitmap   Bitmap of the allocated memory region
+ * @param PhysicalAddress     Pointer to the data part of allocated memory region
+ *
+ * @return EFI_SUCCESS        Successfully allocate the buffer
+ */
+STATIC
+EFI_STATUS
+InternalAllocateBuffer (
+  IN  EFI_ALLOCATE_TYPE        Type,
+  IN  EFI_MEMORY_TYPE          MemoryType,
+  IN  UINTN                    Pages,
+  IN OUT UINT32                *ReservedMemBitmap,
+  IN OUT EFI_PHYSICAL_ADDRESS  *PhysicalAddress
+  )
+{
+  UINT32              MemBitmap;
+  UINT8               Index;
+  RESERVED_MEM_RANGE  *MemRange;
+  UINTN               PagesOfLastMemRange;
+
+  *ReservedMemBitmap = 0;
+
+  if (Pages == 0) {
+    ASSERT (FALSE);
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (!mReservedSharedMemSupported) {
+    goto LegacyAllocateBuffer;
+  }
+
+  if (mReservedSharedMemAddress == 0) {
+    goto LegacyAllocateBuffer;
+  }
+
+  PagesOfLastMemRange = 0;
+
+  for (Index = 0; Index < ARRAY_SIZE (mReservedMemRanges); Index++) {
+    if ((Pages > PagesOfLastMemRange) && (Pages <= EFI_SIZE_TO_PAGES (mReservedMemRanges[Index].DataSize))) {
+      break;
+    }
+
+    PagesOfLastMemRange = EFI_SIZE_TO_PAGES (mReservedMemRanges[Index].DataSize);
+  }
+
+  if (Index == ARRAY_SIZE (mReservedMemRanges)) {
+    // There is no suitable size of reserved memory. Turn to legacy allocate.
+    goto LegacyAllocateBuffer;
+  }
+
+  MemRange = &mReservedMemRanges[Index];
+
+  if ((mReservedMemBitmap & MemRange->BitmapMask) == MemRange->BitmapMask) {
+    // The reserved memory is exausted. Turn to legacy allocate.
+    goto LegacyAllocateBuffer;
+  }
+
+  MemBitmap = (mReservedMemBitmap & MemRange->BitmapMask) >> MemRange->Shift;
+
+  for (Index = 0; Index < MemRange->Slots; Index++) {
+    if ((MemBitmap & (UINT8)(1<<Index)) == 0) {
+      break;
+    }
+  }
+
+  ASSERT (Index != MemRange->Slots);
+
+  *PhysicalAddress   = MemRange->StartAddressOfMemRange + Index * SIZE_OF_MEM_RANGE (MemRange) + MemRange->HeaderSize;
+  *ReservedMemBitmap = (UINT32)(1 << (Index + MemRange->Shift));
+
+  DEBUG ((
+    DEBUG_VERBOSE,
+    "%a: range-size: %lx, start-address=0x%llx, pages=0x%llx, bits=0x%lx, bitmap: %lx => %lx\n",
+    __FUNCTION__,
+    MemRange->DataSize,
+    *PhysicalAddress,
+    Pages,
+    *ReservedMemBitmap,
+    mReservedMemBitmap,
+    mReservedMemBitmap | *ReservedMemBitmap
+    ));
+
+  return EFI_SUCCESS;
+
+LegacyAllocateBuffer:
+
+  *ReservedMemBitmap = 0;
+  return gBS->AllocatePages (Type, MemoryType, Pages, PhysicalAddress);
+}
+
+/**
+ * Allocate reserved shared memory for bounce buffer.
+ *
+ * @param Type        Allocate type
+ * @param MemoryType  The memory type to be allocated
+ * @param MapInfo     Pointer to the MAP_INFO
+ * @return EFI_SUCCESS  Successfully allocate the bounce buffer.

```

Comment:
Is @retval better? @return is usually for the arbitrary return 
description https://sourceforge.net/p/doxygen/mailman/message/11651182/

Reply:
Yes. It will be updated with @RetVal